baseline.gradle

```diff
@@ -60,8 +63,7 @@ subprojects {
     spotless {
       java {
         target 'src/main/java/**/*.java', 'src/test/java/**/*.java', 'src/jmh/java/**/*.java', 'src/integration/java/**/*.java'
-        // we use an older version of google-java-format that is compatible with JDK 8
-        googleJavaFormat("1.7")
+        googleJavaFormat("1.17.0")
```

Comment:
I don't think it's a good idea to upgrade this to the latest version. The issue is that spotless will produce different formatting results depending on which JDK it's been executed with. So as long as the project builds & releases with JDK8 we need to stay at the GJF version 1.7

Reply:
Sooner or later we will add support for Java 21 and we will drop support for Java 8. These two things do not have to depend on each other. Sadly, there is no google java formatter version which can run under Java 21 and under Java 8, so we need to make a choice which Java versions we can check source code formatting with. We can either choose 8-17 or 11-21, but not the union. The 11-21 is an inevitable choice, unless we want to live in the past for ever.
Now, code formatting is very important for project day to day, but feels not directly related to releases. The change here doesn't affect how we build releases.

Reply:
Once Iceberg drops support for Java8 is when we will be able to update the GJF version. As I mentioned earlier, we pinned the version so that `spotlessApply` doesn't produce different formatting results depending on the JDK version it is executed with. Since we release for JDK8, we need to use the GJF version that works with JDK8.

The community is planning to drop JDK8 support (which only Hive depends on) with Iceberg 2.0, which is also when we'll be upgrading most likely to 11-21. That being said, we can only upgrade the GJF version once we also move to JDK 11-21

Reply:
Thanks for rephrasing your initial comment for me. This reinforces my understanding.

>  so that spotlessApply doesn't produce different formatting results depending on the JDK version it is executed with.

I totally agree this wouldn't be acceptable! Do you maybe remember under what circumstances this can be observed?
In this PR I modified the CI scripts so that `./gradlew -DallVersions build -x test -x javadoc -x integrationTest` runs with java 11-21 and all runs agree on the desired formatting.  While I can see that formatting depends on GJF version, I don't see it being dependent on the JDK version it is executed with.

> Since we release for JDK8, we need to use the GJF version that works with JDK8.

I hear you on this, but would appreciate helping me understanding the why. I still fail to see it.
Is it because we expect Iceberg contributors to locally run JDK 8? I am concerned that JDK 8 is unsupported, unmaintained publicly and potentially not safe to run, so would prefer not to force contributors to use it, even though the **releases** target Java 8 compatibility.

Reply:
When I added spotless + configured GJF it would produce different results depending on the JDK version that was used when executing `spotlessApply`. I typically use `sdkman` to switch JDK version and from what I remember is that when you run `spotlessApply` using JDK8 it would produce one formatting and a different formatting when using JDK11+ (which you can also see in this PR, where the formatting changed). Back then I looked into why it behaves the way it does and it was due to  different JDK versions might pick up a different GJF version and so we had to pin the version to the latest one that was supported by JDK8.


Typically whoever does a release is expected to run the release script using JDK8 locally btw. I hope that helps in understanding the context as to why we had to pin the GJF version.

Reply:
thank you for your comment, @nastra !

> using JDK8 it would produce one formatting and a different formatting when using JDK11+ (which you can also see in this PR, where the formatting changed).

in this PR formatting changed, because GJF version changed

>  Back then I looked into why it behaves the way it does and it was due to different JDK versions might pick up a different GJF version and so we had to pin the version to the latest one that was supported by JDK8.

that explains everything, thanks!

Indeed, formatting is apparently dependent on GJF version and GJF version should be pinned. 

on the code from this PR i run the following
```
JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home ./gradlew spotlessApply
JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-17.jdk/Contents/Home ./gradlew spotlessApply
JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-21.jdk/Contents/Home ./gradlew spotlessApply
```

and none of these commands resulted in code changes. The formatting rules seem consistent and remain independent from the JDK version the build runs with.


> Typically whoever does a release is expected to run the release script using JDK8 locally btw.

This is what i was assuming, thanks for confirming this!

Note that with changes here, the build continues to work with JDK 8.
It just doesn't offer code style checks under JDK 8, but the person running release doesn't need them. The release goes off `main` branch where all the checks were already done by the CI. 
If a person running a release wanted to do some code changes before creating the release build -- which sounds wrong to me -- they would still need to re-run all other CI checks, which involve JDK versions 8, 11 and 17 (currently; and 21 should be added to the mix).

> I hope that helps in understanding the context as to why we had to pin the GJF version.

Absolutely! So with this PR, the GJF version remains pinned. I hope this clarifies the concerns.